File: sentinel/src/check-runner.ts
/**
 * Check interface + dispatcher.
 *
 * Each Sentinel check (prompt-injection, pii-leak, secrets-in-input, ...)
 * implements the `Check` interface and gets registered. The dispatcher runs
 * the configured checks against a piece of text and returns any trips.
 *
 * v0.1 ships one stub check (`alwaysTrip`) used in tests and disabled by
 * default. Real checks land Phase 1 W4-5 — they'll plug into this same
 * interface, no changes needed at the wrap() call site.
 */

import type { CheckId, Phase, Severity } from "./types.js";

/** Result from a single check run. */
export interface CheckResult {
  /** Whether this check tripped. False = no incident. */
  tripped: boolean;
  /** Severity, if tripped. Ignored if `tripped: false`. */
  severity: Severity;
  /** Short title (≤80 chars). */
  title: string;
  /** Longer explanation. */
  detail: string;
}

/** A single check. */
export interface Check {
  readonly id: CheckId;
  /** Run the check against a piece of text. Pure if possible — no I/O. */
  run(text: string, phase: Phase): CheckResult | Promise<CheckResult>;
}

/** Result that didn't trip — convenience constant. */
export const PASS: CheckResult = {
  tripped: false,
  severity: "info",
  title: "",
  detail: "",
};

/**
 * In-process registry of available checks. Real checks register here at
 * import time; the wrap() dispatcher looks them up by `CheckId`.
 */
const REGISTRY = new Map<CheckId, Check>();

/** Register a check. Idempotent — re-registering replaces the existing one. */
export function registerCheck(check: Check): void {
  REGISTRY.set(check.id, check);
}

/** Look up a check by id. Returns undefined if not registered. */
export function getCheck(id: CheckId): Check | undefined {
  return REGISTRY.get(id);
}

/** All currently registered check ids. Mainly for diagnostics. */
export function registeredCheckIds(): CheckId[] {
  return Array.from(REGISTRY.keys());
}

/**
 * Run a list of checks against text and return any that tripped.
 * Failures inside a check (thrown errors) are swallowed — Sentinel never
 * blocks a user request because a check has a bug.
 */
export async function runChecks(
  text: string,
  checkIds: readonly CheckId[],
  phase: Phase,
): Promise<Array<{ id: CheckId; result: CheckResult }>> {
  const out: Array<{ id: CheckId; result: CheckResult }> = [];
  for (const id of checkIds) {
    const check = REGISTRY.get(id);
    if (!check) continue; // unknown / not yet implemented; skip silently
    try {
      const result = await check.run(text, phase);
      if (result.tripped) {
        out.push({ id, result });
      }
    } catch {
      // Drop — never block on a check bug.
    }
  }
  return out;
}

// ─── Built-in checks (used in tests; disabled in production) ─────────────────

/**
 * A test-only check that always trips. Registered manually by tests via
 * `registerAlwaysTripCheck()`. NOT auto-registered.
 */
export const ALWAYS_TRIP_ID = "prompt-injection" as const satisfies CheckId;

export const alwaysTripCheck: Check = {
  id: ALWAYS_TRIP_ID,
  run() {
    return {
      tripped: true,
      severity: "high" as Severity,
      title: "always-trip stub fired",
      detail: "This check exists for testing the wrap() lifecycle end-to-end.",
    };
  },
};

File: sentinel-py/src/lictor_sentinel/check_runner.py
"""
Check interface + registry — Python mirror of @lictor/sentinel's
src/check-runner.ts.

Each check implements run(text, phase) -> CheckResult. The registry is
keyed by CheckId; re-registering the same id replaces (used by tests).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Callable, Protocol, runtime_checkable

from .types import CheckId, Phase, Severity


@dataclass(frozen=True)
class CheckResult:
    """Result from a single check run."""

    tripped: bool
    severity: Severity
    title: str
    detail: str


@runtime_checkable
class Check(Protocol):
    """A single Sentinel check."""

    id: CheckId

    def run(self, text: str, phase: Phase) -> CheckResult: ...


PASS = CheckResult(tripped=False, severity="info", title="", detail="")


# ─── Registry ────────────────────────────────────────────────────────────────

_REGISTRY: dict[CheckId, Check] = {}


def register_check(check: Check) -> None:
    """Register a check. Idempotent — re-registering replaces."""
    _REGISTRY[check.id] = check


def get_check(check_id: CheckId) -> Check | None:
    """Look up a check by id."""
    return _REGISTRY.get(check_id)


def registered_check_ids() -> list[CheckId]:
    """All currently registered check ids."""
    return list(_REGISTRY.keys())


def run_checks(
    text: str,
    check_ids: list[CheckId],
    phase: Phase,
) -> list[tuple[CheckId, CheckResult]]:
    """
    Run the given checks against text. Returns trips only.

    Failures inside a check (exceptions) are swallowed — Sentinel never
    blocks a user request because a check has a bug.
    """
    out: list[tuple[CheckId, CheckResult]] = []
    for cid in check_ids:
        check = _REGISTRY.get(cid)
        if check is None:
            continue
        try:
            result = check.run(text, phase)
            if result.tripped:
                out.append((cid, result))
        except Exception:
            # Drop — never block on a check bug.
            pass
    return out


def make_check(
    check_id: CheckId, run_fn: Callable[[str, Phase], CheckResult]
) -> Check:
    """Convenience: build a Check from a function + id."""

    class _AdHocCheck:
        id: CheckId = check_id

        def run(self, text: str, phase: Phase) -> CheckResult:
            return run_fn(text, phase)

    return _AdHocCheck()  # type: ignore[return-value]
